### ISSUE TITLE
option for a <Provider> type like <Loadable.Capture />

### ISSUE THREAD

**Issue by rileylnapier** [faceyspacey/react-universal-component#78]
I think the way I'm using RUC is a little different than the normal way.  I'm SUPER close to having it all work, my problem is that flushChunkNames is always empty.

I think this is because I'm using react-static and creating a "static" react app to render on the server and then also creating a client bundle.   Basically we bundle _everything_ we need into our server bundle used for rendering and doing so creates a copy of react-universal-component in the bundle... then during the render step, i re-import react-universal-component and the 2 global cache can't talk to eachother

so really its the global cache you use to store the chunkNames doesn't persist from my render step and my build step.  This could be remedied  by doing something like what react-loadable does with Loadable.Capture.

```
  let modules = [];

  let html = ReactDOMServer.renderToString(
    <Loadable.Capture report={moduleName => modules.push(moduleName)}>
      <App/>
    </Loadable.Capture>
  );

  console.log(modules);
```

this way we don't need a global cache and I can just pull out the chunk names as i render.

what would be the investment in adding an optional provider type component for smaller use cases like mine???

**Comment by rileylnapier** [comment]
i made this: https://github.com/faceyspacey/react-universal-component/pull/79

**Comment by faceyspacey** [comment]
I can't quite get how the chunk names are lost because of how React-static works?? I no zero about that package.

Also I've tried at all costs to avoid this API surface, but it may be time. I just wanna be sure it's really needed for your scenario and understand why.

**Comment by rileylnapier** [comment]
Well it can be an optional api for people with different setup like me?  This api works while the other doesn’t.  I can create a repo that shows how it doesn’t work.  

I’m not completely sure why it doesn’t... but I imagine what it’s doing is creating a second copy of react-universal-component so we end up with two global cache.  One copy of ruc is in the server bundle and then in a second step we utilized the server bundle to render... we reimport ruc which doesn’t have reference to the global cache of the bundled version?

I also kind of feel like the current setup is a bit “magic” and hard to follow how flushChunkNames gets populated.  I prefer more explicit methods.... but that’s my preference.  


Thoughts?!

**Comment by faceyspacey** [comment]
Using providers scare people. They don't like to add endless provider components to their tree, especially when in this case it's needed in one place (the server) and not on the other (the client). It's more cognitive overhead because now you are wondering what it is doing and perhaps what performance overhead it has on rendering. Now on top of that even though the API is optional it just leaves developers wondering what's the difference and whether they should use it--which is more cognitive overhead for the 80% of users who don't need it.

Still, not against adding it. Just saying.

**Comment by rileylnapier** [comment]
Agreed,  I first ran into this with Stripe’s react components and it confused me as to what was happening.  It was my first introduction with react context....

I tried using react loadable but couldn’t figure it all out... plus they have their “issues” private ~ wtf?!?  The issue archive helps me sooooo many times and helped me understand this library and I turn theirs.

... anyway... I think the problem of apis and providers can be solved with adequate documentation, yeah?  I can write up a section for you and try to explain when to use one and one to use the other???

Also... had been following you on medium fo awhile and have been waiting for an excuse to server render code split app and finally got a project to use it!!! So excited I got things working.  This could be ready big game changer for this ecommerce platform I work for :)

**Comment by rileylnapier** [comment]
hihi!  

i added some readme about the Provider... and i'm in process of making a simple repo using react-static to show how it doesn't work

**Comment by rileylnapier** [comment]
ok, i made a repo:

https://github.com/rileylnapier/react-static-ssr.git

`npm i && npm run build`

universal component used in /src/containers/About.js
![image](https://user-images.githubusercontent.com/7017640/33292568-94c4210e-d37e-11e7-8ebf-26569d645975.png)


renderToString in /static.config.js

![image](https://user-images.githubusercontent.com/7017640/33292584-a24b3452-d37e-11e7-9858-f18a5710c4e1.png)

**Comment by rileylnapier** [comment]
also, if you want to read up on react-static for context: https://medium.com/@tannerlinsley/%EF%B8%8F-introducing-react-static-a-progressive-static-site-framework-for-react-3470d2a51ebc

**Comment by faceyspacey** [comment]
I'm gonna close this out, and shift the conversation to the PR. I'm almost there all with you to bang this out. If not tonight, tomorrow.
